### PR TITLE
Handle Errors when following "next" link when fetching collection

### PR DESCRIPTION
**Pull request by anthonyfalzetti**
While following the next link it times out. Rather than return a partial collection it should raise an error. 

This PR will: 
* Create a new Error type NoSuchLinkError
* When an unsuccessful response is returned raise CollectionError

Example Stacktrace: 
```
** (ExHal.CollectionError) Failed to fetch next link due to timeout
    lib/exhal/collection.ex:24: anonymous fn/1 in ExHal.Collection.to_stream/1
    (elixir) lib/stream.ex:1377: Stream.do_resource/5
    (elixir) lib/stream.ex:1553: Enumerable.Stream.do_each/4
    (elixir) lib/enum.ex:2979: Enum.map/2
```

### CODE REVIEW

File: lib/exhal/collection.ex [hunk 1]
```diff
@@ -14,7 +14,17 @@ defmodule ExHal.Collection do
...
+          # End of Pagination
+          {:error, %ExHal.NoSuchLinkError{}} ->
+            {:halt, nil}
+          {:error, %ExHal.NoSuchLinkError{}, _headers} ->
+            {:halt, nil}
+
+          {:error, %ExHal.Error{reason: msg}} ->
+            raise ExHal.CollectionError, message: "Failed to fetch next link due to #{msg}"
+          {:error, %ExHal.Error{reason: msg}, _headers} ->
+            raise ExHal.CollectionError, message: "Failed to fetch next link due to #{msg}"
```
Comment (heyZeus) [inline lib/exhal/collection.ex:30]: Now `to_stream` raises an exception. Maybe a new function called `to_stream!`

File: lib/exhal/navigation.ex [hunk 2]
```diff
@@ -25,13 +25,13 @@ defmodule ExHal.Navigation do
   @doc """
   Follows all links of a particular rel in a HAL document.
 
-  Returns `[{:ok, %ExHal.Document{...}, %ExHal.ResponseHeader{...}}, {:error, %ExHal.Error{...}, ...]`
+  Returns `[{:ok, %ExHal.Document{...}, %ExHal.ResponseHeader{...}}, {:error, %ExHal.NoSuchLinkError{...}, {:error, %ExHal.Error{...} ...]`
```
Comment (pezra) [inline lib/exhal/navigation.ex]: I wonder if `NoSuchLinkError` should be embedded in a `List`. It is not the same as failing to fetch one target of a compound link. If the document doesn't have a link that's it.

I see a couple of reasonable choices when the link doesn't exist:
 1. return an empty list
 2. return an error tuple like `{:error, %ExHal.NoSuchLinkError{...}}`

If we are going to make a breaking change to the API (which this PR does) we should do one of those things.

File: lib/exhal/collection.ex (hunk 1 above)
Comment (pezra) [reply lib/exhal/collection.ex:30]: I think it makes sense for a stream to raise if there is a problem traversing the list.

**Review by pezra — CHANGES_REQUESTED** [review]
:+1: to raising a error if fetching the next page in a collection fails. However, i don't think this change accomplished the intended effect. I'd like to see some more tests around the target edge case.

File: test/test_helper.exs [hunk 3]
```diff
@@ -3,5 +3,3 @@ Application.ensure_all_started(:mox)
 Application.ensure_all_started(:stream_data)
 
 Mox.defmock(ExHal.ClientMock, for: ExHal.Client)
-
-Application.put_env(:exhal, :client, ExHal.ClientMock)
```
Comment (anthonyfalzetti) [inline test/test_helper.exs:7]: Removed from test_helper.exs since not all tests wish to mock using Mox. Moving Navigation to use runtime client_module this causes issues in testing.

File: test/exhal/form_test.exs [hunk 4]
```diff
@@ -9,6 +9,13 @@ defmodule ExHal.FormTest do
   import Mox
   setup :verify_on_exit!
 
+  setup do
+    Application.put_env(:exhal, :client, ExHal.ClientMock)
+    on_exit fn ->
+      Application.put_env(:exhal, :client, ExHal.Client)
```
Comment (anthonyfalzetti) [inline test/exhal/form_test.exs:15]: When running `mix test test/` if the client is not returned to `ExHal.Client` it fails tests downstream.

File: test/exhal/collection_test.exs [hunk 5]
```diff
@@ -84,6 +86,31 @@ defmodule ExHal.CollectionTest do
     assert ExHal.to_stream(doc) |> is_a_stream
   end
 
+  describe "Non Hal Responses" do
```
Comment (anthonyfalzetti) [inline test/exhal/collection_test.exs:89]: Had to wrap this in a describe block so the setup doesn't affect other tests. Open to a different way to do this.

File: lib/exhal/collection.ex (hunk 1 above)
Comment (anthonyfalzetti) [reply lib/exhal/collection.ex:30]: 👍

File: lib/exhal/navigation.ex (hunk 2 above)
Comment (anthonyfalzetti) [reply lib/exhal/navigation.ex]: I opted to go with `{:error, %ExHal.NoSuchLinkError{...}}` and added test coverage

File: lib/exhal/navigation.ex [hunk 6]
```diff
@@ -1,8 +1,8 @@
 defmodule ExHal.Navigation do
-  @client_module Application.get_env(:exhal, :client, ExHal.Client)
+  # client_module() Application.get_env(:exhal, :client, ExHal.Client)
```
Comment (adherr) [inline lib/exhal/navigation.ex]: Probably want to take this out if it's unused

Comment (anthonyfalzetti) [reply lib/exhal/navigation.ex]: correct. Consider it gone.

File: lib/exhal/collection.ex [hunk 7]
```diff
@@ -14,7 +14,17 @@ defmodule ExHal.Collection do
       fn -> {:ok, a_doc} end,
       fn follow_result ->
         case follow_result do
-          {:error, _} -> {:halt, follow_result}
+          # End of Pagination
```
Comment (alakra) [inline lib/exhal/collection.ex:21]: Since this is now raising, we should probably document that behavior. Here's a proposal for an updated `@doc`:

```
@doc """
Returns a stream that iterates over the collection represented by `a_doc`.

In the case of failures during processing the stream, the stream will halt if 
an `ExHal.NoSuchLinkError` error occurs when following the next resource.  

If an `ExHal.Error` occurs when requesting the next resource, an 
`ExHal.CollectionError` will be raised to the caller.
"""
```

**Review by alakra — APPROVED** [review]

File: lib/exhal/collection.ex (hunk 7 above)
Comment (anthonyfalzetti) [reply lib/exhal/collection.ex:21]: Thanks for the suggestion. Updated.

Comment (adherr) [reply lib/exhal/collection.ex:21]: I think that second paragraph is a little misleading because the `NoSuchLinkError` is not a failure, it's our signal to stop iterating. Maybe add this to the first paragraph?

```
Iteration halts when there is no further `next` link to follow.
```

Comment (alakra) [reply lib/exhal/collection.ex:21]: That's a good point. I second that change.

File: lib/exhal/collection.ex [hunk 8]
```diff
@@ -7,14 +7,28 @@ defmodule ExHal.Collection do
   alias ExHal.ResponseHeader
 
   @doc """
-  Returns a stream that iterate over the collection represented by `a_doc`.
+  Returns a stream that iterates over the collection represented by `a_doc`.
+  Iteration halts when there is no further `next` link to follow.
+
+  If an `ExHal.Error` occurs when requesting the next resource, an
+  `ExHal.CollectionError` will be raised to the caller.
```
Comment (pezra) [inline lib/exhal/collection.ex]: Maybe be a little clearer that a collection error will be raised only when following the next link to the the next page? This currently reads a little like it might raise a collection error if one of the members of the collection is unfetchable.

**Review by pezra — APPROVED** [review]

File: lib/exhal/collection.ex (hunk 8 above)
Comment (comverge-pair) [reply lib/exhal/collection.ex]: updated

Comment (anthonyfalzetti) [reply lib/exhal/collection.ex]: updated

**Comment by anthonyfalzetti** [comment]
Going to merge and release after work today.
